include/boost/serialization/singleton.hpp

```diff
@@ -121,7 +126,8 @@ class singleton : public singleton_module
         // our usage/implementation of "locking" and introduce uncertainty into
         // the sequence of object initializaition.
         use(& m_instance);
-        return static_cast<T &>(t);
+        if (!t) t = new singleton_wrapper;
```

Comment:
This construct does not look thread safe, whereas the original was (at least since C++11). Could this be an issue?

Reply:
Just doing ` static singleton_wrapper *t = new singleton_wrapper;` above should be fine, don't you agree?

Reply:
Hmmm - as far as I'm concerned the original is/was correct.  But it seems that with some combinations of compiler/library ...?  the order of destruction of the static variable is not the reverse of the construction - as they should be.  There is also the question of why we get any kind of memory issue when the program ends. A big motivation for using a static variable was that it would be built into the program image and never touch the heap.  But apparently that was wishful thinking.  It would qualify this as a work around for unexpected/obscure/incorrect runtime behavior regarding invocation of destructor of static variables.  Of course I don't really know.  It could be a bug / undefined behavior which no one has seen yet.  Given all this uncertainty, I'm putting in the change given that it addresses the failure of one of the tests.

On the other hand, I've since changed the test to make it more correct.  It's possible that it was an error in the test that no one sees and that now it's gone. If more information comes to light we can easily change it again.  This is not a persistent object.

So, for now, we'll just keep on trucking ...

Reply:
@masterleinad Yes I think that would fix the concurrency issue I had in mind.

Reply:
About the heap allocation:
This could be avoided by declaring a byte array of singleton_wrapper size (and alignment) and using placement new.

But you'd need `boost::call_once` to ensure thread safety.

While I noticed the problematic order of destruction I have not investigated the reasons in detail, so I unfortunately can't say what exactly goes wrong. But maybe I get around to getting a better explanation.

Reply:
I agree that doing the assignment right in the declaration of the `static` local variable solves the problem. If you don't want to do this, then you need to do double-checking to make things thread-safe, as shown here: https://en.wikipedia.org/wiki/Double-checked_locking

Reply:
Actually, if any of you guys want to make a name for yourself, you could factor out the singleton into it's own boost library/component and the serialization library would just import it.  When I made it it seemed simple - no heap allocation, initialization before main is called.  Destruction after main is exited etc.  A no brainer.  But it might be that the compiler handling of static variables throws a monkey wrench into all this.